### PR TITLE
More Backend Related Refactoring

### PR DESCRIPTION
**Pull request by anstaf**
- native backend is simplified and shared for both strgrid and icgrid versions
- accessor intent enum is suplemented with the utility to apply it.
- `stage` concept for icgrid is tuned a bit  and now it is a superset of `stage` concept for strgrid.
- `intent` is now applied in outside of `iterator_domain`
- the last usage of `CRTP` pattern (`iterator_domain`) is removed from the code base.
- `pointer_oob_check` is removed. The check will be introduced again after refactoring
- `run_functor_arguments` is shrinked

### CODE REVIEW

**Comment by anstaf** [comment]
launch stgrid

**Comment by anstaf** [comment]
launch strgrid

**Comment by anstaf** [comment]
launch icgrid

**Comment by anstaf** [comment]
launch strgrid

**Comment by anstaf** [comment]
launch strgrid

**Comment by anstaf** [comment]
launch icgrid

**Comment by anstaf** [comment]
do perftest

File: include/gridtools/stencil_composition/structured_grids/dim.hpp [hunk 1]
```diff
@@ -17,5 +17,6 @@ namespace gridtools {
         using i = integral_constant<int, 0>;
         using j = integral_constant<int, 1>;
         using k = integral_constant<int, 2>;
+        struct c;
```
Comment (havogt) [inline include/gridtools/stencil_composition/structured_grids/dim.hpp]: What's the reason?

File: include/gridtools/stencil_composition/icosahedral_grids/iterate_domain.hpp [hunk 2]
```diff
@@ -8,124 +8,105 @@
...
-#include "dim.hpp"
 
 namespace gridtools {
-    /**
-       This class is basically the iterate domain. It contains the
-       ways to access data and the implementation of iterating on neighbors.
+    /**@brief class managing the memory accesses, indices increment
+
+       This class gets instantiated in the backend-specific code, and has a different implementation for
+       each backend (see CRTP pattern). It is instantiated within the kernel (e.g. in the device code),
```
Comment (havogt) [inline include/gridtools/stencil_composition/icosahedral_grids/iterate_domain.hpp]: This comment was moved? Please update. (no CRTP?)

File: include/gridtools/stencil_composition/icosahedral_grids/backend_x86/mss_loop_x86.hpp [hunk 3]
```diff
@@ -31,6 +33,28 @@ namespace gridtools {
...
+        template <class LoopIntervals>
+        struct get_ncolors;
+
+        template <template <class...> class L0,
+            template <class...> class L1,
+            template <class...> class L2,
+            class From,
+            class To,
+            class Stage>
+        struct get_ncolors<L0<loop_interval<From, To, L1<L2<Stage>>>>> : Stage::n_colors {};
```
Comment (havogt) [inline include/gridtools/stencil_composition/icosahedral_grids/backend_x86/mss_loop_x86.hpp]: This looks a bit weird. Is there no better way to get the stage?

File: include/gridtools/stencil_composition/backend_naive/iterate_domain_naive.hpp [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        void increment_j(int_t offset = 1) {
+            m_pos.j += offset;
+            increment<dim::j>(offset);
+        }
+        void increment_k(int_t offset = 1) {
+            m_pos.k += offset;
+            increment<dim::k>(offset);
+        }
+
+        void increment_c(int_t offset = 1) { increment<dim::c>(offset); }
```
Comment (havogt) [inline include/gridtools/stencil_composition/backend_naive/iterate_domain_naive.hpp]: Here is the answer to my question about dim::c in structured...

File: include/gridtools/stencil_composition/backend_naive/iterate_domain_naive.hpp [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+            template <class Arg>
+            void operator()() const {
+                auto const &stride = sid::get_stride<Dim>(m_self->template strides<Arg>());
+                auto &ptr = at_key<Arg>(m_self->m_ptr_map);
+                sid::shift(ptr, stride, m_offset);
+            }
+        };
+
+        template <class Grid>
+        struct set_base_offset_f {
```
Comment (havogt) [inline include/gridtools/stencil_composition/backend_naive/iterate_domain_naive.hpp]: Why is this base offset backend specific?

File: include/gridtools/stencil_composition/backend_naive/fused_mss_loop_naive.hpp [hunk 6]
```diff
@@ -9,34 +9,94 @@
...
+                int_t k_from = m_grid.template value_at<From>();
+                int_t k_to = m_grid.template value_at<To>();
+                int_t i_count = m_grid.i_high_bound() - m_grid.i_low_bound() + 1 + iplus - iminus;
+                int_t j_count = m_grid.j_high_bound() - m_grid.j_low_bound() + 1 + jplus - jminus;
+                int_t k_count = 1 + std::abs(k_to - k_from);
+                int_t k_step = k_to >= k_from ? 1 : -1;
+
+                iterate_domain_naive<LocalDomain> it_domain(m_local_domain, m_grid);
+
+                it_domain.increment_i(iminus);
```
Comment (havogt) [inline include/gridtools/stencil_composition/backend_naive/fused_mss_loop_naive.hpp]: I think some comments would be useful here. The naive implementation can then serve as an example for new backend developers.

**Comment by havogt** [comment]
Anton, can you try to limit the size of this refactoring PRs? I don't know, maybe set yourself a hard limit on lines changes. It is very hard to follow which code is moved and which is new.

File: include/gridtools/stencil_composition/icosahedral_grids/backend_cuda/iterate_domain_cuda.hpp [hunk 7]
```diff
@@ -9,61 +9,89 @@
...
+
+        using caches_t = typename IterateDomainArguments::local_domain_t::cache_sequence_t;
+        using ij_cache_args_t = GT_META_CALL(ij_cache_args, caches_t);
+        using k_cache_args_t = GT_META_CALL(k_cache_args, caches_t);
 
         using readwrite_args_t = GT_META_CALL(compute_readwrite_args, typename IterateDomainArguments::esf_sequence_t);
-        using iterate_domain_cache_t = iterate_domain_cache<IterateDomainArguments>;
+
+        // array storing the (i,j) position of the current thread within the block
+        array<int, 2> m_thread_pos;
```
[the remark this answers is not in the document]
Comment (anstaf) [reply include/gridtools/stencil_composition/icosahedral_grids/backend_cuda/iterate_domain_cuda.hpp]: done

**Comment by havogt** [comment]
Needs update of reference (sync with #1281).

File: include/gridtools/stencil_composition/icosahedral_grids/iterate_domain.hpp (hunk 2 above)
Comment (anstaf) [reply include/gridtools/stencil_composition/icosahedral_grids/iterate_domain.hpp]: done

File: include/gridtools/stencil_composition/icosahedral_grids/backend_x86/mss_loop_x86.hpp (hunk 3 above)
Comment (anstaf) [reply include/gridtools/stencil_composition/icosahedral_grids/backend_x86/mss_loop_x86.hpp]: I add a comment their

File: include/gridtools/stencil_composition/backend_naive/iterate_domain_naive.hpp (hunk 5 above)
Comment (anstaf) [reply include/gridtools/stencil_composition/backend_naive/iterate_domain_naive.hpp]: It is not. This code is moved here temporary and eventually it will gone . The perpose is to make naive backend as much stand alone as possible.

File: include/gridtools/stencil_composition/backend_naive/fused_mss_loop_naive.hpp (hunk 6 above)
Comment (anstaf) [reply include/gridtools/stencil_composition/backend_naive/fused_mss_loop_naive.hpp]: done

**Comment by anstaf** [comment]
launch strgrid

**Comment by anstaf** [comment]
launch icgrid

**Comment by anstaf** [comment]
do perftest
